### ISSUE TITLE
Customized loss function need to know the train/val split

### ISSUE THREAD

**Issue by ouyangzhibo** [KevinMusgrave/powerful-benchmarker#76]
Hi, my customized loss function needs to load a file that is dependent on the train/val partition. So is it possible to let the customized loss function know which train/val partition it is trained on? How can I achieve that? Many thanks for the help!

**Comment by KevinMusgrave** [comment]
Here's how you can pass the current train/val split into your loss function

```python
from powerful_benchmarker.api_parsers import BaseAPIParser
from powerful_benchmarker.factories import LossFactory
from powerful_benchmarker.utils import common_functions as c_f
from pytorch_metric_learning import losses
import copy

class CustomAPIParser(BaseAPIParser):
    def default_kwargs_loss_funcs(self):
        kwargs = super().default_kwargs_loss_funcs()
        kwargs["split_manager"] = lambda: self.split_manager
        return kwargs

class CustomLossFactory(LossFactory):
    def _create_general(self, loss_type, num_classes, split_manager):

        ############################################
        #### all of this is from the base class ####
        ############################################
        loss, loss_params = self.getter.get("loss", yaml_dict=loss_type, return_uninitialized=True)
        loss_params = copy.deepcopy(loss_params)
        if loss == losses.MultipleLosses:
            loss_funcs = [self.create({k:v}) for k,v in loss_params["losses"].items()]
            return loss(loss_funcs) 

        if loss == losses.CrossBatchMemory:
            if "loss" in loss_params: 
                loss_params["loss"] = self.create(loss_params["loss"])
            if "miner" in loss_params: 
                self.miner_factory = self.miner_factory(getter=self.getter)
                loss_params["miner"] = self.miner_factory.create(loss_params["miner"])

        if c_f.check_init_arguments(loss, "num_classes") and ("num_classes" not in loss_params):
            loss_params["num_classes"] = num_classes
            logging.info("Passing %d as num_classes to the loss function"%loss_params["num_classes"])

        for obj_name in self.nested_loss_objects:
            if obj_name in loss_params:
                if "regularizer" in obj_name: 
                    loss_params[obj_name] = self.get_nested_regularizer(loss_params[obj_name])
                elif obj_name == "weight_init_func":
                    loss_params[obj_name] = self.get_weight_init_func(loss_params[obj_name])
                else:
                    loss_params[obj_name] = self.getter.get(obj_name, yaml_dict=loss_params[obj_name])
        ############################################
        ####    end of stuff from base class    ####
        ############################################

        loss_params["train_set"] = split_manager.get_dataset("eval", "train")
        loss_params["val_set"] = split_manager.get_dataset("eval", "val")

        return loss(**loss_params)
```

The important part is at the bottom of ```CustomLossFactory```. The current training set and validation set are retrieved using ```split_manager.get_dataset```. The arguments to this function are ```transform_type``` and ```split_name```. So if you want to get the training set using the training transform, you can do ```split_manager.get_dataset("train", "train")```.

Next, you need to tell the benchmarker to use the custom api parser and custom loss factory:
```
python run.py \
--api_parser {CustomAPIParser: {}} \
--factories {loss~OVERRIDE~: {CustomLossFactory: {}}}
```

Let me know if that works. The above code was tested with v0.9.33.

**Comment by ouyangzhibo** [comment]
Hi @KevinMusgrave, many thanks for the solution! But I might not have made question clear (sorry about that). So there are 4 different partitions (labeled as p0-p3) in total using leave-on-out experiment. I would like to know how to let the loss function know which partition (p0, p1, p2 or p3) it is trained on at the loss function initialization stage. Let me know if it is clearer.

**Comment by ouyangzhibo** [comment]
> Hi @KevinMusgrave, many thanks for the solution! But I might not have made question clear (sorry about that). So there are 4 different partitions (labeled as p0-p3) in total using leave-on-out experiment. I would like to know how to let the loss function know which partition (p0, p1, p2 or p3) it is trained on at the loss function initialization stage. Let me know if it is clearer.

I have figured out how to add it from your example code. There is a property `curr_split_scheme_name` in the `split_manager ` and that is exactly what I want. Many thanks for the help!
